File: day6/ans.py
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_map = defaultdict(list)
        for s in strs:
            anagram_map[self.eleCount(s)].append(s)
        
        return anagram_map.values()
    
    def eleCount(self, s):
        count = [0]*26
        for c in s:
            count[ord(c) - ord('a')] += 1
        
        return tuple(count)

File: day6/ans.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        std::unordered_map<std::string, std::vector<std::string>> anagram_map;
        for(auto& s: strs) {
            anagram_map[countingSort(s)].push_back(s);
        }
        
        std::vector<std::vector<string>> ans;
        ans.reserve(anagram_map.size());
        for(auto& kv: anagram_map) {
            ans.push_back(std::move(kv.second));
        }
        
        return ans;
    }

private:
    string countingSort(string& s) {
        char count[26] = {0};
        for(auto c: s) {
            count[c - 'a'] += 1;
        }
        
        string sort_s;
        for(int i = 0; i < 26; i++) {
            sort_s.append(count[i], 'a' + i);
        }
        return sort_s;
    }
};
